File: data/skills/scrnaseq-seurat-core-analysis/scripts/remove_ambient_rna.R
#' Ambient RNA Correction for Single-Cell RNA-seq Data
#'
#' This module provides functions for removing ambient RNA contamination using SoupX.
#'
#' For detailed guidance, see references/ambient_rna_correction.md

library(Seurat)
library(SoupX)
library(ggplot2)
library(ggprism)


#' Run SoupX ambient RNA correction
#'
#' SoupX estimates and removes ambient RNA contamination from droplet-based
#' scRNA-seq data by comparing raw and filtered count matrices.
#'
#' @param raw_matrix_dir Path to raw_feature_bc_matrix/ directory
#' @param filtered_matrix_dir Path to filtered_feature_bc_matrix/ directory
#' @param clusters Optional cluster assignments for cells (improves estimation)
#' @param output_dir Directory to save SoupX outputs
#' @param auto_estimate Auto-estimate contamination fraction (default: TRUE)
#' @param contamination_range Range for contamination estimation (default: c(0.01, 0.5))
#'
#' @return Seurat object with corrected counts
#'
#' @export
run_soupx_correction <- function(
  raw_matrix_dir,
  filtered_matrix_dir,
  clusters = NULL,
  output_dir = "results/soupx",
  auto_estimate = TRUE,
  contamination_range = c(0.01, 0.5)
) {
  # Create output directory
  dir.create(output_dir, showWarnings = FALSE, recursive = TRUE)

  # Load raw and filtered matrices
  cat("Loading raw and filtered matrices...\n")
  raw_data <- Read10X(raw_matrix_dir)
  filtered_data <- Read10X(filtered_matrix_dir)

  # Create Seurat object from filtered data for clustering if needed
  if (is.null(clusters)) {
    cat("Performing quick clustering for contamination estimation...\n")
    seurat_temp <- CreateSeuratObject(counts = filtered_data)
    seurat_temp <- NormalizeData(seurat_temp, verbose = FALSE)
    seurat_temp <- FindVariableFeatures(seurat_temp, verbose = FALSE)
    seurat_temp <- ScaleData(seurat_temp, verbose = FALSE)
    seurat_temp <- RunPCA(seurat_temp, verbose = FALSE)
    seurat_temp <- FindNeighbors(seurat_temp, dims = 1:30, verbose = FALSE)
    seurat_temp <- FindClusters(seurat_temp, resolution = 0.8, verbose = FALSE)
    clusters <- Idents(seurat_temp)
  }

  # Create SoupChannel
  sc <- SoupChannel(raw_data, filtered_data)

  # Add cluster information
  sc <- setClusters(sc, clusters)

  # Estimate contamination
  if (auto_estimate) {
    cat("Auto-estimating contamination fraction...\n")
    sc <- autoEstCont(sc)
    contamination_fraction <- sc$metaData$rho[1]
    cat(sprintf("Estimated contamination fraction: %.3f (%.1f%%)\n",
                contamination_fraction, contamination_fraction * 100))
  } else {
    # Manual estimation
    cat("Manually estimating contamination fraction...\n")
    sc <- estimateContamination(sc, tfidfMin = 0.5, verbose = TRUE)
    contamination_fraction <- sc$fit$rho
  }

  # Correct counts
  cat("Correcting counts...\n")
  corrected_counts <- adjustCounts(sc)

  # Create Seurat object with corrected counts
  seurat_corrected <- CreateSeuratObject(counts = corrected_counts)

  # Store metadata
  seurat_corrected@misc$soupx <- list(
    contamination_fraction = contamination_fraction,
    raw_matrix_dir = raw_matrix_dir,
    filtered_matrix_dir = filtered_matrix_dir
  )

  # Save output
  saveRDS(seurat_corrected, file = file.path(output_dir, "seurat_soupx_corrected.rds"))
  write.csv(data.frame(contamination_fraction = contamination_fraction),
            file = file.path(output_dir, "contamination_estimate.csv"),
            row.names = FALSE)

  cat("SoupX correction completed successfully\n")
  return(seurat_corrected)
}


#' Estimate soup (ambient RNA) fraction
#'
#' Extract contamination fraction from SoupX-corrected Seurat object
#'
#' @param seurat_obj Seurat object with SoupX metadata
#'
#' @return Contamination fraction (0-1)
#'
#' @export
estimate_soup_fraction <- function(seurat_obj) {
  if (is.null(seurat_obj@misc$soupx)) {
    warning("No SoupX metadata found in Seurat object")
    return(0.0)
  }
  return(seurat_obj@misc$soupx$contamination_fraction)
}


#' Plot SoupX correction results
#'
#' Visualize the impact of ambient RNA correction
#'
#' @param seurat_obj Seurat object with SoupX metadata
#' @param output_dir Directory to save plots
#'
#' @export
plot_soupx_results <- function(seurat_obj, output_dir = "results/soupx") {
  dir.create(output_dir, showWarnings = FALSE, recursive = TRUE)

  contamination_fraction <- estimate_soup_fraction(seurat_obj)

  # Plot contamination estimate
  p <- ggplot(data.frame(value = contamination_fraction), aes(x = 1, y = value)) +
    geom_col(fill = "#4575b4", width = 0.5) +
    ylim(0, 0.5) +
    labs(
      title = "Estimated Ambient RNA Contamination",
      y = "Contamination Fraction",
      x = ""
    ) +
    theme_prism() +
    theme(axis.text.x = element_blank(), axis.ticks.x = element_blank())

  ggsave(file.path(output_dir, "contamination_estimate.svg"),
         plot = p, width = 4, height = 6, dpi = 300)

  cat(sprintf("Contamination fraction: %.1f%%\n", contamination_fraction * 100))
}


#' Compare counts before and after SoupX correction
#'
#' @param seurat_before Seurat object before correction
#' @param seurat_after Seurat object after correction
#' @param marker_genes Vector of marker gene names to compare
#' @param output_dir Directory to save comparison plots
#'
#' @export
compare_before_after_soupx <- function(
  seurat_before,
  seurat_after,
  marker_genes = NULL,
  output_dir = "results/soupx"
) {
  dir.create(output_dir, showWarnings = FALSE, recursive = TRUE)

  # Compare total counts per cell
  before_counts <- colSums(seurat_before@assays$RNA@counts)
  after_counts <- colSums(seurat_after@assays$RNA@counts)

  comparison_df <- data.frame(
    before = before_counts,
    after = after_counts
  )

  # Scatter plot
  p_scatter <- ggplot(comparison_df, aes(x = before, y = after)) +
    geom_point(alpha = 0.3, size = 0.5, color = "#4575b4") +
    geom_abline(slope = 1, intercept = 0, linetype = "dashed", color = "red") +
    labs(
      title = "Total Counts Before vs After Correction",
      x = "Before Correction",
      y = "After Correction"
    ) +
    theme_prism()

  ggsave(file.path(output_dir, "counts_comparison.svg"),
         plot = p_scatter, width = 6, height = 6, dpi = 300)

  # Summary statistics
  reduction_pct <- (1 - mean(after_counts) / mean(before_counts)) * 100

  cat(sprintf("Mean counts before: %.1f\n", mean(before_counts)))
  cat(sprintf("Mean counts after: %.1f\n", mean(after_counts)))
  cat(sprintf("Reduction: %.1f%%\n", reduction_pct))

  # If marker genes provided, compare their expression
  if (!is.null(marker_genes)) {
    marker_genes <- intersect(marker_genes, rownames(seurat_before))

    if (length(marker_genes) > 0) {
      before_expr <- as.matrix(seurat_before@assays$RNA@counts[marker_genes, ])
      after_expr <- as.matrix(seurat_after@assays$RNA@counts[marker_genes, ])

      marker_comparison <- data.frame(
        gene = rep(marker_genes, each = ncol(seurat_before)),
        before = as.vector(t(before_expr)),
        after = as.vector(t(after_expr))
      )

      p_markers <- ggplot(marker_comparison, aes(x = before, y = after)) +
        geom_point(alpha = 0.3, size = 0.5) +
        geom_abline(slope = 1, intercept = 0, linetype = "dashed", color = "red") +
        facet_wrap(~gene, scales = "free") +
        labs(
          title = "Marker Gene Expression Before vs After",
          x = "Before Correction",
          y = "After Correction"
        ) +
        theme_prism()

      ggsave(file.path(output_dir, "marker_comparison.svg"),
             plot = p_markers, width = 12, height = 8, dpi = 300)
    }
  }
}


#' Run SoupX with custom parameters for high-soup tissues
#'
#' Specialized function for tissues with high ambient RNA (brain, lung, tumor)
#'
#' @param raw_matrix_dir Path to raw_feature_bc_matrix/ directory
#' @param filtered_matrix_dir Path to filtered_feature_bc_matrix/ directory
#' @param output_dir Directory to save outputs
#'
#' @return Seurat object with corrected counts
#'
#' @export
run_soupx_high_soup <- function(
  raw_matrix_dir,
  filtered_matrix_dir,
  output_dir = "results/soupx"
) {
  # Use more aggressive settings for high-soup tissues
  seurat_corrected <- run_soupx_correction(
    raw_matrix_dir = raw_matrix_dir,
    filtered_matrix_dir = filtered_matrix_dir,
    output_dir = output_dir,
    auto_estimate = TRUE,
    contamination_range = c(0.05, 0.5)  # Higher expected contamination
  )

  contamination <- estimate_soup_fraction(seurat_corrected)

  if (contamination > 0.15) {
    cat("WARNING: High contamination detected (>15%).\n")
    cat("Consider reviewing filtering thresholds and data quality.\n")
  }

  return(seurat_corrected)
}


# Example usage
if (FALSE) {
  # Run SoupX correction
  seurat_corrected <- run_soupx_correction(
    raw_matrix_dir = "raw_feature_bc_matrix/",
    filtered_matrix_dir = "filtered_feature_bc_matrix/",
    output_dir = "results/soupx"
  )

  # Check contamination
  contamination <- estimate_soup_fraction(seurat_corrected)
  cat(sprintf("Contamination: %.1f%%\n", contamination * 100))

  # Plot results
  plot_soupx_results(seurat_corrected, output_dir = "results/soupx")
}

File: data/skills/scrnaseq-scanpy-core-analysis/scripts/remove_ambient_rna.py
"""
Ambient RNA Correction for Single-Cell RNA-seq Data

This module provides functions for removing ambient RNA contamination using
CellBender or SoupX-based approaches.

For detailed guidance, see references/ambient_rna_correction.md
"""

import scanpy as sc
import numpy as np
import pandas as pd
import subprocess
import os
from pathlib import Path
from typing import Union, Optional, Tuple
import warnings


def run_cellbender(
    raw_h5: Union[str, Path],
    expected_cells: int,
    total_droplets: Optional[int] = None,
    output_dir: Union[str, Path] = "results/cellbender",
    epochs: int = 200,
    fpr: float = 0.01,
    use_cuda: bool = True,
    **kwargs
) -> sc.AnnData:
    """
    Run CellBender remove-background to correct ambient RNA.

    CellBender uses a deep generative model to estimate and remove ambient RNA
    contamination from droplet-based scRNA-seq data.

    Parameters
    ----------
    raw_h5 : str or Path
        Path to raw_feature_bc_matrix.h5 file from CellRanger
    expected_cells : int
        Expected number of real cells in the dataset
    total_droplets : int, optional
        Total droplets to include (should be 2-3x expected_cells)
        If None, automatically set to 3x expected_cells
    output_dir : str or Path
        Directory to save CellBender outputs
    epochs : int
        Number of training epochs (default: 200)
    fpr : float
        False positive rate for cell calling (default: 0.01)
    use_cuda : bool
        Use GPU acceleration if available (default: True)
    **kwargs
        Additional arguments passed to cellbender remove-background

    Returns
    -------
    adata : AnnData
        Corrected count matrix as AnnData object

    Notes
    -----
    - Requires CellBender installation: pip install cellbender
    - GPU recommended for speed (10-20x faster than CPU)
    - For high-soup tissues (brain, lung, tumor), set total_droplets = 3x expected_cells
    """
    # Set up paths
    raw_h5 = Path(raw_h5)
    output_dir = Path(output_dir)
    output_dir.mkdir(parents=True, exist_ok=True)

    if total_droplets is None:
        total_droplets = expected_cells * 3

    output_h5 = output_dir / "cellbender_output.h5"

    # Build command
    cmd = [
        "cellbender", "remove-background",
        "--input", str(raw_h5),
        "--output", str(output_h5),
        "--expected-cells", str(expected_cells),
        "--total-droplets-included", str(total_droplets),
        "--epochs", str(epochs),
        "--fpr", str(fpr)
    ]

    if use_cuda:
        cmd.append("--cuda")

    # Add any additional arguments
    for key, value in kwargs.items():
        cmd.extend([f"--{key.replace('_', '-')}", str(value)])

    print(f"Running CellBender with command:")
    print(" ".join(cmd))

    try:
        result = subprocess.run(cmd, check=True, capture_output=True, text=True)
        print("CellBender completed successfully")
    except subprocess.CalledProcessError as e:
        print(f"CellBender failed with error:\n{e.stderr}")
        raise

    # Load corrected counts
    adata = sc.read_10x_h5(str(output_h5).replace(".h5", "_filtered.h5"))
    adata.var_names_make_unique()

    # Store metadata
    adata.uns["cellbender"] = {
        "raw_h5": str(raw_h5),
        "expected_cells": expected_cells,
        "total_droplets": total_droplets,
        "epochs": epochs,
        "fpr": fpr
    }

    return adata


def run_soupx_python(
    raw_matrix_dir: Union[str, Path],
    filtered_matrix_dir: Union[str, Path],
    clusters: Optional[np.ndarray] = None,
    output_dir: Union[str, Path] = "results/soupx"
) -> sc.AnnData:
    """
    Run SoupX-based ambient RNA correction using Python/scanpy.

    This is a simplified Python implementation of the SoupX algorithm
    for removing ambient RNA contamination.

    Parameters
    ----------
    raw_matrix_dir : str or Path
        Path to raw_feature_bc_matrix/ directory
    filtered_matrix_dir : str or Path
        Path to filtered_feature_bc_matrix/ directory
    clusters : array-like, optional
        Pre-computed cluster labels for cells
        If None, quick clustering will be performed
    output_dir : str or Path
        Directory to save outputs

    Returns
    -------
    adata : AnnData
        Corrected count matrix as AnnData object

    Notes
    -----
    This is a simplified implementation. For full SoupX functionality,
    use the R version via rpy2 or run R script directly.
    """
    # Load raw and filtered matrices
    adata_raw = sc.read_10x_mtx(raw_matrix_dir)
    adata_filtered = sc.read_10x_mtx(filtered_matrix_dir)

    # Estimate ambient profile from empty droplets
    empty_cells = adata_raw.obs_names[~adata_raw.obs_names.isin(adata_filtered.obs_names)]
    ambient_profile = adata_raw[empty_cells, :].X.sum(axis=0).A1
    ambient_profile = ambient_profile / ambient_profile.sum()

    # Quick clustering if not provided
    if clusters is None:
        sc.pp.normalize_total(adata_filtered, target_sum=1e4)
        sc.pp.log1p(adata_filtered)
        sc.pp.highly_variable_genes(adata_filtered, n_top_genes=2000)
        sc.pp.pca(adata_filtered, n_comps=30)
        sc.pp.neighbors(adata_filtered, n_neighbors=15)
        sc.tl.leiden(adata_filtered, resolution=0.8)
        clusters = adata_filtered.obs["leiden"].values

    # Estimate contamination fraction
    contamination_fraction = estimate_contamination_fraction(
        adata_filtered.X.toarray(),
        ambient_profile,
        clusters
    )

    # Correct counts
    corrected = adata_filtered.X.toarray() - contamination_fraction * ambient_profile
    corrected = np.maximum(corrected, 0)  # No negative counts

    # Create corrected AnnData
    adata_corrected = sc.AnnData(
        X=corrected,
        obs=adata_filtered.obs,
        var=adata_filtered.var
    )

    # Store metadata
    adata_corrected.uns["soupx"] = {
        "contamination_fraction": contamination_fraction,
        "ambient_profile": ambient_profile
    }

    # Save output
    output_dir = Path(output_dir)
    output_dir.mkdir(parents=True, exist_ok=True)
    adata_corrected.write_h5ad(output_dir / "soupx_corrected.h5ad")

    return adata_corrected


def estimate_contamination_fraction(
    counts: np.ndarray,
    ambient_profile: np.ndarray,
    clusters: np.ndarray
) -> float:
    """
    Estimate global contamination fraction.

    Parameters
    ----------
    counts : array-like, shape (n_cells, n_genes)
        Raw count matrix
    ambient_profile : array-like, shape (n_genes,)
        Ambient RNA profile from empty droplets
    clusters : array-like, shape (n_cells,)
        Cluster assignments

    Returns
    -------
    rho : float
        Estimated contamination fraction (0-1)
    """
    # Use marker genes with high specificity
    # Simple heuristic: genes expressed in <10% of clusters
    n_clusters = len(np.unique(clusters))
    cluster_expression = np.zeros((n_clusters, counts.shape[1]))

    for i, cluster in enumerate(np.unique(clusters)):
        cluster_mask = clusters == cluster
        cluster_expression[i, :] = (counts[cluster_mask, :] > 0).mean(axis=0)

    # Marker genes: expressed in few clusters
    marker_mask = (cluster_expression > 0.1).sum(axis=0) <= n_clusters * 0.1

    if marker_mask.sum() < 10:
        # Fallback: use all genes
        marker_mask = np.ones(counts.shape[1], dtype=bool)
        warnings.warn("Few marker genes found, using all genes for contamination estimation")

    # Estimate contamination as correlation between observed and ambient
    observed_profile = counts.mean(axis=0)

    # Correlation-based estimation
    marker_obs = observed_profile[marker_mask]
    marker_amb = ambient_profile[marker_mask]

    # Estimate rho
    rho = np.corrcoef(marker_obs, marker_amb)[0, 1]
    rho = np.clip(rho, 0, 0.5)  # Reasonable range

    return rho


def estimate_contamination(adata: sc.AnnData) -> float:
    """
    Estimate ambient RNA contamination fraction from corrected AnnData.

    Parameters
    ----------
    adata : AnnData
        AnnData object with ambient correction metadata

    Returns
    -------
    contamination : float
        Estimated contamination fraction
    """
    if "cellbender" in adata.uns:
        # Extract from CellBender metadata if available
        # This is approximate - CellBender doesn't directly report this
        return 0.05  # Placeholder
    elif "soupx" in adata.uns:
        return adata.uns["soupx"]["contamination_fraction"]
    else:
        warnings.warn("No ambient correction metadata found")
        return 0.0


def compare_before_after(
    adata_before: sc.AnnData,
    adata_after: sc.AnnData,
    marker_genes: list,
    output_dir: Union[str, Path] = "results/ambient"
):
    """
    Compare expression before and after ambient RNA correction.

    Parameters
    ----------
    adata_before : AnnData
        Data before correction
    adata_after : AnnData
        Data after correction
    marker_genes : list
        List of marker genes to compare
    output_dir : str or Path
        Directory to save comparison plots
    """
    from plotnine import ggplot, aes, geom_point, labs, theme_minimal, facet_wrap
    from plotnine_prism import theme_prism
    import matplotlib.pyplot as plt

    output_dir = Path(output_dir)
    output_dir.mkdir(parents=True, exist_ok=True)

    # Compare total counts
    before_counts = adata_before.X.sum(axis=1).A1
    after_counts = adata_after.X.sum(axis=1).A1

    comparison_df = pd.DataFrame({
        "before": before_counts,
        "after": after_counts
    })

    plot = (
        ggplot(comparison_df, aes(x="before", y="after"))
        + geom_point(alpha=0.3, size=0.5)
        + labs(
            title="Total Counts Before vs After Correction",
            x="Before Correction",
            y="After Correction"
        )
        + theme_prism()
    )

    plot.save(output_dir / "counts_comparison.svg", dpi=300, width=6, height=6)

    print(f"Mean counts before: {before_counts.mean():.1f}")
    print(f"Mean counts after: {after_counts.mean():.1f}")
    print(f"Reduction: {(1 - after_counts.mean()/before_counts.mean()) * 100:.1f}%")


# Example usage
if __name__ == "__main__":
    # Example: Run CellBender
    adata = run_cellbender(
        raw_h5="raw_feature_bc_matrix.h5",
        expected_cells=10000,
        output_dir="results/cellbender"
    )

    print(f"Loaded corrected data: {adata.shape[0]} cells, {adata.shape[1]} genes")
